File: asg4.pl
%(* /*
%* CSCI3180 Principles of Programming Languages
% *
% * --- Declaration ---
%*
% * I declare that the assignment here submitted is original except for source
% * material explicitly acknowledged. I also acknowledge that I am aware of
% * University policy and regulations on honesty in academic work, and of the
% * disciplinary guidelines and procedures applicable to breaches of such policy
% * and regulations, as contained in the website
% * http://www.cuhk.edu.hk/policy/academichonesty/
% *
% * Assignment 4
% * Name : Rustami Ubaydullo
% * Student ID : 1155102622
% * Email Addr : 1155102622@link.cuhk.edu.hk
% */ *)


sum(0,X,X).
sum(s(X),Y,s(Z)) :- sum(X,Y,Z).

% 1.a Rules
natNum(0).
natNum(s(X)) :- natNum(X).

% 1.b Rules
lt(0,s(_)).
lt(s(X),s(Y)) :- lt(X,Y).

% 1.c Query
% ?- lt(X,s(s(s(0)))).

% 1.d Rules
geq(T, T).
geq(_,0).
geq(X,Y) :- lt(Y,X).

% 1.e Rules
max(X,Y,Z) :- geq(X,Y), X=Z.
max(X,Y,Z) :- geq(Y,X), Y=Z.

% 1.f Rules
difference(X,Y,Z) :- sum(Z,Y,X), geq(X,Y).

% 1.g Rules Credit: Huzeyfe Kiran
mod(X,Y,Z) :- lt(Z,Y), geq(X,Z), difference(X,Z,D), multiple(Y,D).
multiple(X,0) :- X=s(_).
multiple(X,Y) :- X=s(_), Y=s(_), geq(Y,X), difference(Y,X,D), multiple(X,D).

% 2.a Rules
% bt(a,bt(b,nil,bt(d,nil,nil)),bt(c,bt(e,bt(f,nil,nil),bt(g,nil,nil)),nil)).

% 2.b Rules
isTree(nil).
isTree(bt(_,L,R)) :- isTree(L), isTree(R).

% 2.c Rules
isLeaf(E,bt(E,nil,nil)).
isLeaf(E,bt(Rt,L,R)) :- isTree(bt(Rt,L,R)),isLeaf(E,L).
isLeaf(E,bt(Rt,L,R)) :- isTree(bt(Rt,L,R)),isLeaf(E,R).

% 2.d Rules Credit: Huzeyfe Kiran
numberOfLeaf(nil,N) :- N=0.
numberOfLeaf(bt(E,nil,nil),N) :- isTree(bt(E,nil,nil)), N=s(0).
numberOfLeaf(bt(Rt,L,R),N) :- isTree(bt(Rt,L,R)), numberOfLeaf(L,LN), numberOfLeaf(R,RN), sum(LN,RN,N).

% 2.e Rules Credit: Huzeyfe Kiran
height(nil,N) :- N=0.
height(bt(_Rt,nil,nil),N) :- N=s(0).
height(bt(Rt,L,R),N) :- isTree(bt(Rt,L,R)), height(L,LH), height(R,RH), max(RH,LH,M), sum(s(0),M,N).

File: asg4.ml
(* /*
 * CSCI3180 Principles of Programming Languages
 *
 * --- Declaration ---
 *
 * I declare that the assignment here submitted is original except for source
 * material explicitly acknowledged. I also acknowledge that I am aware of
 * University policy and regulations on honesty in academic work, and of the
 * disciplinary guidelines and procedures applicable to breaches of such policy
 * and regulations, as contained in the website
 * http://www.cuhk.edu.hk/policy/academichonesty/
 *
 * Assignment 4
 * Name : Rustami Ubaydullo
 * Student ID : 1155102622
 * Email Addr : 1155102622@link.cuhk.edu.hk
 */ *)

datatype suit = Clubs | Diamonds | Hearts | Spades;
datatype rank = Jack | Queen | King | Ace | Num of int;
type card = suit * rank;
datatype color = Red | Black;
datatype move = Discard of card | Draw;
fun abs x = if x >= 0 then x else ~x;

(* Credit: Huzeyfe Kiran *)
fun cardColor(x:suit,y:rank):color = if x=Spades orelse x=Clubs then Black else Red;
fun cardValue(_,Jack) = 10
|   cardValue(_,Queen) = 10
|   cardValue(_,Ace) = 11
|   cardValue(_,King) = 10
|   cardValue(_,Num i) = i;

fun numOfCard(x:(suit*rank) list):int = length x;
fun removeCard(l:(suit*rank) list, s:suit, r:rank):(suit*rank) list =
        if (s,r) = hd(l) then tl(l)
        else [hd(l)]@removeCard(tl(l),s,r);

fun sumCards(l:(suit*rank) list, c:color):int =
        if null(l) then 0
        else if cardColor(hd(l)) = c then cardValue(hd(l)) + sumCards(tl(l),c)
        else sumCards(tl(l),c);

fun score(l:(suit*rank) list) = abs(sumCards(l,Red) - sumCards(l,Black)) + (6 - length(l));

fun runGame(deck,mvList) =
            let
                    fun runGameA([],Draw,mvList,hand) = score(hand)
                    |   runGameA(deck,Draw,mvList,hand) =
                                if length(hd(deck)::hand) > 5 orelse null(mvList) then score(hd(deck)::hand)
                                else runGameA(tl(deck),hd(mvList),tl(mvList),hd(deck)::hand)
                    |   runGameA(deck,Discard (x,y),mvList,hand) =
                                if null(mvList) then score(removeCard(hand,x,y))
                                else runGameA(deck,hd(mvList),tl(mvList),removeCard(hand,x,y))

            in
                    runGameA(deck,hd(mvList),tl(mvList),[])
            end;
